pkg/pgmodel/pgx.go

```diff
@@ -647,112 +647,68 @@ func runCopyFrom(conn pgxConn, in chan copyRequest) {
-
 // we can currently recover from two error:
 // If we inserted duplicate data we switch to using INSERT ... ON CONFLICT DO NOTHING
 // to skip redundant data points and try again.
 // If we inserted into a compressed chunk, we decompress the chunk and try again.
 // Since a single batch can have both errors, we need to remember the insert method
 // we're using, so that we deduplicate if needed.
-func tryRecovery(conn pgxConn, req copyRequest, insertFn insertionFunction, err error) (insertionFunction, error) {
-
+func tryRecovery(conn pgxConn, req copyRequest, err error) error {
 	// we only recover from postgres errors right now
 	pgErr, ok := err.(*pgconn.PgError)
 	if !ok {
 		errMsg := err.Error()
 		log.Warn("msg", fmt.Sprintf("unexpected error while inserting to %s", req.table), "error", errMsg)
-		return nil, err
-	}
-
-	// if we have duplicate data points try again using
-	// INSERT ... ON CONFLICT DO NOTHING to insert only the new ones
-	if pgErr.Code == pgerrcode.UniqueViolation {
-		log.Warn("msg", fmt.Sprintf("duplicate data in sample for %s", req.table), "detail", pgErr.Detail, "table", req.table)
-		req.data.batch.ResetPosition()
-		return doInsert, nil
+		return err
 	}
 
 	// If the error was that the table is already compressed, decompress and try again.
 	if strings.Contains(pgErr.Message, "insert/update/delete not permitted") {
 		decompressErr := decompressChunks(conn, req.data, req.table)
 		if decompressErr != nil {
-			return nil, err
+			return err
 		}
 
 		req.data.batch.ResetPosition()
-		// use the old insertFn in case it was doing deduplication
-		return insertFn, nil
+		return nil
 	}
 
 	log.Warn("msg", fmt.Sprintf("unexpected postgres error while inserting to %s", req.table), "error", pgErr.Error())
-	return nil, err
-}
-
-func doCopyFrom(conn pgxConn, req copyRequest) error {
-	_, err := conn.CopyFrom(
-		context.Background(),
-		pgx.Identifier{dataSchema, req.table},
-		copyColumns,
-		&req.data.batch,
-	)
 	return err
 }
 
 func doInsert(conn pgxConn, req copyRequest) (err error) {
-	queryString := fmt.Sprintf("INSERT INTO %s(time, value, series_id) VALUES ($1, $2, $3) ON CONFLICT DO NOTHING", pgx.Identifier{dataSchema, req.table}.Sanitize())
-	insertBatch := conn.NewBatch()
-	calls := int64(0)
-	for req.data.batch.Next() {
-		values, _ := req.data.batch.Values()
-		insertBatch.Queue(queryString, values...)
-		calls++
+	numRows := 0
+	for i := range req.data.batch.sampleInfos {
+		numRows += len(req.data.batch.sampleInfos[i].samples)
 	}
-	var resp pgx.BatchResults
-	resp, err = conn.SendBatch(context.Background(), insertBatch)
+	times := make([]time.Time, 0, numRows)
+	vals := make([]float64, 0, numRows)
+	series := make([]int64, 0, numRows)
+	for req.data.batch.Next() {
+		row, _ := req.data.batch.Values()
+		times = append(times, row[0].(time.Time))
+		vals = append(vals, row[1].(float64))
+		series = append(series, int64(row[2].(SeriesID)))
+	}
```

Comment:
can we check that len(times) == numRows? and throw an error if not aligned?

Reply:
it'll be a panic but sure. (It's a panic because `numRows` is gathered by iterating through the same slices; if there's an error it's a non-recoverable bug)